### PR TITLE
Create OAuthClient for OpenShift OAuth Proxy

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
Use OAuthClient per Workspace to get user:full token with OpenShift OAuth Proxy.

Drawbacks:
- currently, oauthclient secrets are hard-coded but there is a separate issue where this part should be addressed along with cookie-secret;
- Since WorkspaceRouting are namespaced-scoped and OAuthClient are cluster-scoped, WorkspaceRouting can't own OAuthClient. So, I added finalizer and OAuth clients are removed along with workspaces but when OAuthClient are removed or updated - the reconciling loop is not reinvoked. Probably we're able to listen to all OAuthClients as we do with an object owned by us (Workspace,WorkspaceRouting), check if it's workspace-related OAuthClient and reinvoke reconcile loop for WorkspaceRouting. Not sure if we really should implement it. Seems Che Operator does not watch it's oauthclient as well.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16515

### Is it tested? How?
oauth-client

### CODE REVIEW

File: pkg/controller/workspacerouting/sync_oauthclients.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	return found.Items, nil
+}
+
+func (r *ReconcileWorkspaceRouting) deleteOAuthClients(routing *v1alpha1.WorkspaceRouting) error {
+	labelSelector, err := labels.Parse(fmt.Sprintf("%s=%s", config.WorkspaceIDLabel, routing.Spec.WorkspaceId))
+	if err != nil {
+		return err
+	}
+	listOptions := &client.DeleteAllOfOptions{
+		ListOptions: *&client.ListOptions{
```
Comment (amisevsk) [inline pkg/controller/workspacerouting/sync_oauthclients.go:110]: Looks like some IDE "magic": `*&client.ListOptions`

File: pkg/controller/workspacerouting/workspacerouting_controller.go [hunk 2]
```diff
@@ -195,9 +230,28 @@ func (r *ReconcileWorkspaceRouting) Reconcile(request reconcile.Request) (reconc
 		return reconcile.Result{}, statusErr
 	}
 
+	oauthClientInSync, err := r.syncOAuthClient(instance, *oauthClient)
```
Comment (amisevsk) [inline pkg/controller/workspacerouting/workspacerouting_controller.go:233]: This looks like a potential nil dereference -- `routingObjects.OAuthClient` should be nil for `basic` routing.

File: pkg/controller/workspacerouting/workspacerouting_controller.go [hunk 3]
```diff
@@ -245,3 +299,21 @@ func getSecureProtocol(protocol string) string {
...
+}
+
+func remove(list []string, s string) []string {
+	for i, v := range list {
+		if v == s {
+			list = append(list[:i], list[i+1:]...)
+		}
+	}
+	return list
+}
```
Comment (amisevsk) [inline pkg/controller/workspacerouting/workspacerouting_controller.go:319]: Gotta love golang.

File: pkg/controller/workspacerouting/sync_oauthclients.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func (r *ReconcileWorkspaceRouting) getClusterOAuthClients(routing *v1alpha1.WorkspaceRouting) ([]oauthv1.OAuthClient, error) {
+	found := &oauthv1.OAuthClientList{}
+	labelSelector, err := labels.Parse(fmt.Sprintf("%s=%s", config.WorkspaceIDLabel, routing.Spec.WorkspaceId))
+	if err != nil {
+		return nil, err
+	}
+	listOptions := &client.ListOptions{
+		Namespace:     routing.Namespace,
```
Comment (amisevsk) [inline pkg/controller/workspacerouting/sync_oauthclients.go:93]: I had issues with the namespaced list options here, since OAuthClients are cluster wide.

**Comment by amisevsk** [comment]
Fixed the issues I mentioned and rebased onto master to resolve conflict.

File: Makefile [hunk 5]
```diff
@@ -100,6 +100,7 @@ _update_controller_configmap:
 
 _apply_controller_cfg:
 	$(TOOL) apply -f ./deploy
+	$(TOOL) apply -f ./deploy/role.yaml
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply Makefile:103]: Screenshot_20200422_145654
It's related ) For some reason oauthclients disappear from role when applying the whole folder. And I have a plan to remove applying the whole folder in a separate PR.

**Pull request merged**